src/routes/publish.js

```diff
@@ -42,29 +45,78 @@ export function publishRoutes() {
 
     const hasherTransform = new HasherPassThrough();
     const lengthChecker = new LengthCheckerPassThrough(MAX_BODY_LENGTH);
+    // The payload should "look like" a json object.
+    const jsonChecker = new CheapJsonChecker();
+    const gunzipStream = new GunzipWrapper();
     const concatener = new Concatenator();
 
     // The "pipeline" utility conveniently destroys all streams when there's an
     // error. However the HTTP request isn't part of the pipeline because we
     // don't want to destroy it. Indeed if it's destroyed, the socket is closed
     // and we can't send a nice error to the caller.
-    const pipeline = util.promisify(Stream.pipeline);
+    const pipeline = promisify(Stream.pipeline);
+    const finished = promisify(Stream.finished);
+
+    // We create 2 interconnected pipelines:
+    //
+    //     request
+    //        |
+    //  lengthChecker --- gunzipStream --- jsonChecker
+    //        |
+    // hasherTransform
+    //        |
+    //   concatener
+    //
+    // These streams' possible errors are handled partly by nodejs' pipeline
+    // function, and some special error handlers to forward errors between
+    // pipelines.
+
+    // 1. First and main pipeline
     const pipelinePromise = pipeline(
       lengthChecker,
       hasherTransform,
       concatener
     );
 
+    // 2. "Gunzip and check" json pipeline
+    // This pipeline will stop quite early, once we know this looks like a json.
+    // When it ends we'll just unpipe.
+    // We don't use the `pipeline` utility because with it we get some
+    // "premature close" errors when the length checker errors out.
+    lengthChecker.pipe(gunzipStream);
+    gunzipStream.pipe(jsonChecker);
+    const jsonCheckerPromise = finished(jsonChecker);
+
+    // Stop the flow by unpiping when we're done checking the JSON.
+    jsonChecker.on('finish', () => {
+      log.verbose(
+        'json-checker-pipeline-done',
+        'The stream pipeline to check the json content is finished.'
+      );
+
+      // Note this is important to unpipe, because jsonChecker stopped consuming
+      // the data, and so if we don't stop the flow, the main pipeline will also
+      // stop flowing once the buffers in the second pipeline are full.
+      lengthChecker.unpipe(gunzipStream);
+      gunzipStream.destroy();
+    });
+
+    // Forward errors between streams in the second pipeline, as well as the
+    // first pipeline through lengthChecker.
+    // Note that errors inside the first pipeline are forwarded thanks to the
+    // `pipeline` thingie.
+    forwardErrors(lengthChecker, gunzipStream, jsonChecker);
```

Comment:
Wow, this feels like magic. (To be clear, because the streams API is magical)

Reply:
Forwardum Errorus